### ISSUE TITLE
Any limits on Git LFS

### ISSUE THREAD

**Issue by shanehsu** [MicrosoftDocs/azure-devops-docs#4179]
Are there any limits on single file size or total storage size for Git LFS? 

---
#### 文件詳細資料

⚠ *請勿編輯此區段。 這是 docs.microsoft.com ➟ GitHub 問題連結的必要項。*

* ID: 96e60bfd-765f-81ad-312d-f46ef9a66ddc
* Version Independent ID: c7ea62ca-1ba4-9a63-30d5-4788f4dce0c4
* Content: [Git limits - Azure Repos](https://docs.microsoft.com/zh-tw/azure/devops/repos/git/limits?view=azure-devops#feedback)
* Content Source: [docs/repos/git/limits.md](https://github.com/MicrosoftDocs/vsts-docs/blob/master/docs/repos/git/limits.md)
* Product: **devops**
* Technology: **devops-code-git**
* GitHub Login: @steved0x
* Microsoft Alias: **sdanie**

**Comment by shanehsu** [comment]
Can you guys get back to me here when you got a reply from the dev team? Putting a feature on hold for this.

**Comment by bluewwork** [comment]
tutoril pls,with vid in windows 10

**Comment by macjustice** [comment]
It appears that LFS is also affected by the 5GB push limit. I'm opening a support request to see if there's any way to increase that limit for LFS usage.

**Comment by arammaliachi** [comment]
@macjustice thank you for spotting this, the change has been submitted and is under review by the Content Development team. In short:

> Git LFS pushes don’t count towards the 5GB push limit. There is no limit for the size of the files in LFS. The 5GB limit is only for files in the actual repo, not blobs stored as part of LFS.

**Comment by ssa3512** [comment]
@arammaliachi I am not able to push a 342 MB file with `git lfs push`. Azure devops returns a HTTP 413 (Request Entity Too Large).

**Comment by ssa3512** [comment]
I was able to determine a ~120MB file size limit using this method:

- Create new azure devops git repository
- Run the following powershell in an emtpy folder
```
git init
git lfs install
git remote add origin [remote url]
git lfs track *.bin
git add -A
git commit -m "Initial Commit"

for ($i = 10; $i -le 350; $i += 10) {
    Write-Host $i
    fsutil file createnew $("testfile_" + $i + "MB.bin") ($i * 1024 * 1024)
    git add -A
    git commit -m "Added $i MB file"
    git push origin master
}
```
This generates arbitrary files in increasing 10MB steps. A 120MB file will push fine, but a 130MB file and anything larger fails with an HTTP 413 error.

This seems like a crazy low limit for Git LFS as GitHub supports files up to 2GB.

**Comment by arammaliachi** [comment]
@ssa3512 **HTTP 413** is usually a grumpy proxy or any other networking/machine limitation. I have seen tickets for that. Please [open up a support request for the Azure DevOps support team](https://azure.microsoft.com/en-us/support/devops/) and send the ID to my email: v-armal@microsoft.com

**Comment by shanehsu** [comment]
@macjustice Sorry to bother, but is there any update on increasing the total push size beyond 5 GB? Also, I'd like to know if there is a soft or hard limit on **total LFS object size**.

**Comment by arammaliachi** [comment]
@shanehsu product group is not currently considering increasing the 5GB limit/push operation for Git repositories that **do not** implement LFS. The 5GB limit is hard for files that are not tracked by LFS.

I don't think we have a size limit per LFS object.

**Comment by macjustice** [comment]
Hello, in my testing I was able to upload multiple files above the 5GB
threshold using LFS. I haven’t run into a limit on total LFS object
storage, either, but I haven’t pushed it that hard yet.

**Comment by arammaliachi** [comment]
@shanehsu; @macjustice; @ssa3512; I think we have a size limit per LFS object of 128MBs, will confirm and update the documentation accordingly.

**Comment by arammaliachi** [comment]
@shanehsu ; @macjustice ; @ssa3512 Running the following command within the affected repository may get you unblocked:

`git config http.version HTTP/1.1`

Then retry the push.

**Comment by ssa3512** [comment]
@arammaliachi that worked for me. I was able to push the file I couldn’t previously.

**Comment by dshakya** [comment]
I am trying to push a file over 5GB with LFS. It starts to push using http/1.1 but restarts after approximately 2 mins. On a faster upload connection, i get upto 1.4Gb and on slow upload speed only get upto about 112MB and then restarts. The time is about 2 mins on both speed. Any help is appreciated. I can see the file being listed using git lfs ls-files

**Comment by dshakya** [comment]
I did a test by creating a VM on Azure. Clone the repo and added the 5GB file and pushed using faster connection. This time, it went above 3GB and then crashed at 2 mins mark. (Again!) Is there something I need to tweak so that the push does not restart at 2 mins?

05:18:03.288870 trace git-lfs: tq: starting transfer adapter "basic"
05:18:03.289855 trace git-lfs: HTTP: PUT https://<DOMAIN>.visualstudio.com/<FOLDER>/info/lfs/objects/918f62a074bef6d1f3ffc4117ee226093bad7d0ce3bec8d7735fc3682923841e
05:20:08.434420 trace git-lfs: HTTP: 503B | 19 MB/s
05:20:08.436423 trace git-lfs: tq: retrying object 918f62a074bef6d1f3ffc4117ee226093bad7d0ce3bec8d7735fc3682923841e: Fatal error: Server error: https://<DOMAIN>.visualstudio.com/<FOLDER>/info/lfs/objects/918f62a074bef6d1f3ffc4117ee226093bad7d0ce3bec8d7735fc3682923841e from HTTP 503
05:20:08.436423 trace git-lfs: tq: enqueue retry #1 for "918f62a074bef6d1f3ffc4117ee226093bad7d0ce3bec8d7735fc3682923841e" (size: 5402327976)
05:20:08.436423 trace git-lfs: tq: sending batch of size 1
05:20:08.436423 trace git-lfs: api: batch 1 files

........AFTER MULTIPLE RESTART......

Uploading LFS objects:   0% (0/1), 8.7 GB | 34 MB/s

**Comment by JSalazar88** [comment]
> @shanehsu ; @macjustice ; @ssa3512 Running the following command within the affected repository may get you unblocked:
> 
> `git config http.version HTTP/1.1`
> 
> Then retry the push.

This appears to work for us as well

**Comment by norlin** [comment]
Any confirmed answer on this?
It's weird that there is no info about how much data I can store in total with git+lfs with Azure

**Comment by tedvanderveen** [comment]
Same problem here. Mostly HTTP 413 errors and sometimes 503 error. Relatively small number of files (35) total size around 250Mb. Nothing too special.

**Comment by moyukhbiswas** [comment]
> @shanehsu ; @macjustice ; @ssa3512 Running the following command within the affected repository may get you unblocked:
> 
> `git config http.version HTTP/1.1`
> 
> Then retry the push.

I was facing the same issue and this works. Also, there seems to be more cases of this failure from my geolocation (India), so that might be something to note.

**Comment by pri-kise** [comment]
Is there any progress or news? We need to know if there are any limits on single file size or total storage size for Git LFS, to decide if we are going to use Azure DevOps with Git LFS.

**Comment by ssa3512** [comment]
> Is there any progress or news? We need to know if there are any limits on single file size or total storage size for Git LFS, to decide if we are going to use Azure DevOps with Git LFS.

There are no limits for Git LFS, however there is a bug surrounding HTTP/2 that requires you to change a git config setting on any client that needs to push files larger than a certain threshold (seeemingly around 120MB)

**Comment by MattJeanes** [comment]
The `git config http.version HTTP/1.1` fix worked for us too uploaded a blob of around 800MB, original error response was `HTTP/2.0 413 Request Entity Too Large` using `GIT_CURL_VERBOSE=1` and `GIT_TRACE=1`

**Comment by pri-kise** [comment]
Is anyone else having the problem that the upload restarts after some time?
Is there any timeout ?

**Comment by kwaclaw** [comment]
Yes git config http.version HTTP/1.1 is the solution.
My uploads failed at about 140MB.

**Comment by FlsZen** [comment]
This still seems to be an issue, luckily `git config http.version HTTP/1.1` is still the workaround.

**Comment by tknerr** [comment]
Same issue for me as @dshakya reported above in https://github.com/MicrosoftDocs/azure-devops-docs/issues/4179#issuecomment-588619424 : trying to upload a larger LFS file (also with `git config http.version HTTP/1.1`) reproducibly times out after 2 minutes + 5 seconds due to an HTTP 503 returned by the server.

I tried fiddling with the various git-lfs timeout settings but to no avail. It always restarts the transfer after 2m 5s due to an HTTP 503...

@arammaliachi is this a known issue?

**Comment by jwittner** [comment]
Following up to report I also had a case of HTTP 413 repeatedly for a ~220mb file before the magic of `git config http.version HTTP/1.1` was applied and the upload worked.

**Comment by RonPeters** [comment]
I had the HTTP 503 problem with a 166MB file. `git config http.version HTTP/1.1` did the trick for me. But since you can't automatically include such configuration in every user's git config, I think Microsoft needs to fix their HTTP2 proxy for a long-term solution.

**Comment by FezVrasta** [comment]
👋 may someone please confirm there aren't any limits on how many GB of LFS files can I store on my repository? It seems too good to be true.

**Comment by sldr** [comment]
> 👋 may someone please confirm there aren't any limits on how many GB of LFS files can I store on my repository? It seems too good to be true.

Of course their are limits even if MS said "no limits". I know of no ISD (Infinite Storage Drive), yet 😉.

I think what you are looking for is [here](https://docs.microsoft.com/en-us/azure/devops/repos/git/limits?view=azure-devops) (https://docs.microsoft.com/en-us/azure/devops/repos/git/limits?view=azure-devops).

BTW: Ran into the problem too n git config... allowed it to work.

**Comment by Sidewinder53** [comment]
Since forcing the HTTP version to 1.1 seems to be the solution (or workaround) for this issue, can we please get the docs updated to reflect this?

This has also been confirmed in another thread over at the Developer Community: https://developercommunity.visualstudio.com/t/cant-push-large-git-lfs-files-to-repository-from-p/862165#T-N1154018

**Comment by NuLL3rr0r** [comment]
This is an annoying known issue. In fact, I [wrote an article on it](https://www.babaei.net/blog/host-unreal-engine-projects-microsoft-azure-devops-unlimited-cost-free-git-lfs-quota/), in order to make hosting Unreal Engine projects, work on Azure DevOps. Now, I did an engine upgrade, and have a Git Commit of around 166 MB and 42 GB of new LFS content. The LFS content has been pushed just fine. The issue for me is now I'm not able to push the commit itself. The interesting part was the initial version of the engine I pushed in one commit was much larger than this 166MB commit.

> **UPDATE 8 [2023/03/04]**: _After upgrading my project to Unreal Engin 5.1 despite the fact that I've already set the git configuration <code>http.version</code> to <code>HTTP/1.1</code> as instructed in this article, despite the commit size of no bigger than <code>166.30 MB</code> and the acceptable upload bandwidth I've got, I was getting <code>HTTP 413 Request Entity Too Large</code> error:_

```
Enumerating objects: 190058, done.
Counting objects: 100% (164439/164439), done.
Delta compression using up to 16 threads
Compressing objects: 100% (113439/113439), done.
Writing objects: 100% (138834/138834), 166.30 MiB | 47.32 MiB/s, done.
Total 138834 (delta 35613), reused 121343 (delta 22206), pack-reused 0
error: RPC failed; HTTP 413 curl 22 The requested URL returned error: 413
send-pack: unexpected disconnect while reading sideband packet
fatal: the remote end hung up unexpectedly
Everything up-to-date
```

> _I tried every suggestion that I came across in order to debug and resolve the issue to no avail. Including enabling git verbose logging:_

```
$ export GIT_TRACE_PACKET=1                      
$ export GIT_TRACE=1
$ export GIT_CURL_VERBOSE=1
```

> _And then maxing out on all the size limits, buffer, and packet sizes, and other hints:_


```
[core]
	compression = 0
    packedGitLimit = 512m
    packedGitWindowSize = 512m
[http]
    postBuffer = 2147483648
[https]
    postBuffer = 2147483648
[init]
    defaultBranch = master
[pack] 
    deltaCacheSize = 2047m
    packSizeLimit = 2047m
	window = 1
    windowMemory = 2047m
```

_Then I tried to change the origin URL to SSH:_

```
[remote "origin"]
	url = https://SOME-ORGANIZATION@dev.azure.com/SOME-ORGANIZATION/MamadouArchives/_git/MamadouArchives
	#url = git@ssh.dev.azure.com:v3/SOME-ORGANIZATION/MamadouArchives/MamadouArchives
```

_And then pushing without LFS:_

```
$ git push --set-upstream origin 5.1 --no-verify
```

> This attempt was futile as well, that made me revert back to https. Then I tried to push commit by commit since I had made a few commits using (<code>5.1</code> which has been repeated twice in the following command, is the name of the new local branch intended to be pushed):


```
$ git rev-list --reverse 5.1 | ruby -ne 'i ||= 0; i += 1; puts $_ if i % 1 == 0' | xargs -I{} git push origin +{}:refs/heads/5.1 --no-verify
```

> _And sadly, the approach of pushing one commit at a time was frutiless as well :/_

**Comment by NuLL3rr0r** [comment]
Solved the issue by creating a new organization and a new repository inside it. This repository is no way near the 10 GB limit as the git count object shows:
```
$ git count-objects -vH                         
count: 0               
size: 0 bytes
in-pack: 296583
packs: 1
size-pack: 366.76 MiB
prune-packable: 0
garbage: 0
size-garbage: 0 bytes
```

There might be some organization-level limit for Azure DevOps I personally am not aware of. See the link from the previous post for the details.
